[17.0] [MIG] stock_putaway_hook: Migration to 17.0

alexandregaldeano opened the pull request:
* replaced `fields_view_get` to `_get_view` and within it, `attrs` updates have been replaced by direct attributes changes
* added `odoo_test_helper` as test dependency

Commit:
Add stock_putaway_hook

Add hooks allowing modules to add more putaway strategies.
It adds hooks to the core putaway method
``StockLocation._get_putaway_strategy()`` allowing to plug other
strategies and makes the selector fields in the tree views dynamic
(required/readonly).

The fields_view_get override is to handle a tricky issue:

By default, the product_id field as such attrs:

"{'readonly': [('category_id', '!=', False)], 'required': [('category_id', '=', False)]}"

And the category_id field:

"{'readonly': [('product_id', '!=', False)], 'required': [('product_id', '=', False)]}"

If we add a new field "route_id", we want to extend the domains to
include it, such as:

"{'readonly': ['|', ('category_id', '!=', False), ('route_id '!=', False)],
  'required': [('category_id', '=', False), ('route_id', '=', False)]}"

And every new field will need to be added in the product_id and
category_id field, but also in the other fields (exemple, if we add a
"foo_id" field, it needs to be added in the attrs of "route_id"), which
isn't possible with the extension mechanisms of the XML views.

The solution provided by this module is that new fields must have the
{'exclusive_selection': True} in the options attribute, and an override
of fields_view_get automatically generates the attrs for all these
"exclusive" fields.

The code here will be tested by `stock_putaway_by_route` in a follow-up
commit.

OCA-git-bot committed:
[ADD] icon.png

Commit:
Add match by product's route in putaway by routes

The module matching a putaway rule by a route can now:

* find a putaway rule by the route related to a stock move
* or a route related to the product on the move

The changes required to enable this feature:

* In stock_putaway_hook: in the part that match the value, support "in"
  when the value is a container (recordset, list, tuple), so we can pass
  product.route_ids (and match any route set on a put-away rule)
* In stock_putaway_by_route, in the various parts where
  `StockLocation._get_putaway_strategy()` is called, propagate the
  product's routes as well
* In StockMoveLine.onchange_product_id, remove the computation based on
  self.move_id: it cannot change.

OCA-git-bot committed:
stock_putaway_hook 13.0.1.1.0

hparfr committed:
[MIG] stock_putaway_hook: Migration to 14.0

hparfr committed:
stock_putaway_rule: adapt to new odoo api

See https://github.com/odoo/odoo/commit/aae1d578296875a641d1142c0f73279fc134a151

flachica committed:
[MIG] stock_putaway_hook: 15.0

rousseldenis committed:
[MIG] stock_putaway_hook: Migration to 16.0

rousseldenis committed:
[FIX] stock_putaway_hook: The _get_putaway_strategy() should always return a location           mer 05 oct 2022 12:41:58

From v15, the _get_putaway_strategy() method on stock locations shoul always return a location.

rousseldenis committed:
[IMP] stock_putaway_hook: Add tests

Commit:
[UPD] Update stock_putaway_hook.pot

OCA-git-bot committed:
[UPD] README.rst

OCA-git-bot committed:
[UPD] README.rst

pedrobaeza committed:
[FIX] *: Linter fixes

OCA-git-bot committed:
[BOT] post-merge updates

mymage committed:
Added translation using Weblate (Italian)

mymage committed:
Translated using Weblate (Italian)

Currently translated at 100.0% (2 of 2 strings)

Translation: stock-logistics-workflow-16.0/stock-logistics-workflow-16.0-stock_putaway_hook
Translate-URL: https://translation.odoo-community.org/projects/stock-logistics-workflow-16-0/stock-logistics-workflow-16-0-stock_putaway_hook/it/

alexandregaldeano committed:
[IMP] stock_putaway_hook: pre-commit auto fixes

alexandregaldeano committed:
[MIG] stock_putaway_hook: Migration to 17.0

rousseldenis commented:
/ocabot migration stock_putaway_hook

rousseldenis commented:
@alexandregaldeano Thanks for this. A test with `Form` could be great as `_get_view()` override is always touchy. Thanks

alexandregaldeano commented:
@rousseldenis that was a great suggestion, I understand more how this code works. I have added tests as you suggested. But now I'm questioning whether the _get_view override is useful at all. I have added the following test to check what was the behavior without the module: https://github.com/OCA/stock-logistics-workflow/pull/1842/commits/95a9784efc3cee014b82e6e660690650d7f5a2f9#diff-0c5186a1fcedfbbd1cd6b4d951d6a372b6c527330313759f55edb1f9e4a4f0a7R74
I feel like the changes in how Odoo 17 works make this part of the change irrelevant, what do you think?

sebalix reviewed (commented):
Thanks @alexandregaldeano for this migration

About the `_get_view` override if it is still relevant: it is if we still want to migrate `stock_putaway_by_route` that was adding a `route_id` field with the `exclusive_selection` option set.

However @jbaudoux thinks this was a bad design to configure route on put-away rules, and therefore we don't plan to use `stock_putaway_by_route` in the future. That means the current override of `_get_view` is not useful anymore as it was done for this purpose initially.

My take: we should drop all the `_get_view` plumbery (less code to maintain).

cc @jbaudoux

jbaudoux commented:
> However @jbaudoux thinks this was a bad design to configure route on put-away rules, and therefore we don't plan to use `stock_putaway_by_route` in the future. That means the current override of `_get_view` is not useful anymore as it was done for this purpose initially.

`stock_putaway_by_route` wasn't a bad design but a simple solution I did in v10.0 and that was migrated up to v14.0. Now, odoo integrated the package type concept for put-away and there is the `stock_storage_type` module for advanced routing. So there is no reason to continue to support `stock_putaway_by_route`.

It is dropped in the v18.0 migration. You could also cherry-pick https://github.com/OCA/stock-logistics-workflow/pull/1931/commits/8c3ca0151f4a1f339cbd6cde476141fcb91735d5 here

alexandregaldeano commented:
@jbaudoux I integrated changes from https://github.com/OCA/stock-logistics-workflow/commit/8c3ca0151f4a1f339cbd6cde476141fcb91735d5 into the migration commit if that's okay to you?
If not I could add an additional commit just before the [pre-commit auto fixes](https://github.com/OCA/stock-logistics-workflow/pull/1842/commits/3c84227210151f5636fa7b75316f8d858ebc417c) one :)

jbaudoux commented:
> @jbaudoux I integrated changes from [8c3ca01](https://github.com/OCA/stock-logistics-workflow/commit/8c3ca0151f4a1f339cbd6cde476141fcb91735d5) into the migration commit if that's okay to you? If not I could add an additional commit just before the [pre-commit auto fixes](https://github.com/OCA/stock-logistics-workflow/pull/1842/commits/3c84227210151f5636fa7b75316f8d858ebc417c) one :)

Thanks

sebalix commented:
Thanks @alexandregaldeano , I'll update https://github.com/OCA/stock-logistics-workflow/pull/1931 accordingly.

jbaudoux commented:
/ocabot merge nobump
